packages/apputils/src/sessioncontext.tsx

```diff
@@ -1484,21 +1510,241 @@ export class SessionContextDialogs implements ISessionContext.IDialogs {
   private _settingRegistry: ISettingRegistry | null;
 }
 
+export namespace SessionContextDialogs {
+  /**
+   * Populate a kernel select node for the session.
+   *
+   * #### Notes
+   * If `options.preference.language` is set, the dropdown is populated with the
+   * following option groups and options:
+   *
+   *  - (Start %1 Kernel, language)
+   *    - { all kernelspecs whose language matches in alphabetical order }
+   *  - (Use No Kernel)
+   *    - `No Kernel`
+   *  - (Start Kernel)
+   *    - { all other kernelspecs in alphabetical order }
+   *  - (Connect to Existing %1 Kernel, language)
+   *    - { all running kernels whose language matches in alphabetical order }
+   *  - (Connect to Kernel)
+   *    - { all other running kernels in alphabetical order }
+   *
+   * If no language preference is set the dropdown is populated with:
+   *
+   *  - (Start Kernel)
+   *    - { all kernelspecs in alphabetical order }
+   *  - (Use No Kernel)
+   *    - `No Kernel`
+   *  - (Connect to Existing Kernel)
+   *    - { all running kernels in alphabetical order  }
+   *
+   * If `options.preference.id` is set and a kernel exists with that id, its
+   * corresponding `<option>` is selected. Otherwise if
+   * `options.preference.language` is set, the first kernelspec that matches it
+   * is selected.
+   */
+  export function populateKernelSelect(
```

Comment:
As this aims to be public, it would be more useful to return the list of options (probably grouped, something like `{ [group: string]: KernelOption[] }`). Like that if someone wants to define its own flavor of the dialog, it can reuse this and we don't expose a function with an interface too close to the UI implementation details.

Reply:
Fair enough. I think because a `<select>` tag is a known quantity it's not so bad. But I agree with your design principle here. The only reason I exposed it was for testing, but even for this reason, it's nicer to taste data validity than to use DOM methods to query values. I'll update this today.